### PR TITLE
feat(kubernetes): checks for memory limits, memory requests, and image tag

### LINKED ISSUE #7630: Missing documented workload checks in prowler kubernetes scan output
### New feature motivation

Hi Prowler team 👋,

According to the official documentation on the Prowler website ([Kubernetes Policy Index](https://docs.prowler.com/checks/kubernetes/kubernetes-policy-index/)), there are several valuable workload-level checks listed, such as:

- Ensure liveness probe is configured
- Ensure readiness probe is configured
- Ensure CPU request is set
- Ensure CPU limits are set
- Ensure memory requests are set
- Ensure memory limits are set
- Ensure image tag is set to Fixed – not Latest or Blank

However, when I run prowler kubernetes --list-checks using version v5.5.1, none of these checks appear in the output. Additionally, even after deploying intentionally misconfigured workloads (e.g., missing probes or resource limits), they do not show up in the scan results.

### Solution Proposed

These are very important security and reliability best practices, especially for production Kubernetes environments, and it would be great to have them included by default in the scan.

### Describe alternatives you've considered

-

### Additional context

Environment:

Prowler version: v5.5.1
Mode: Kubernetes scan running in-cluster
Platform: EKS

Would appreciate any clarification, and happy to help test if needed. Thanks for your great work on this project! 🙌

### PR DESCRIPTION
**Pull request by Nikhilkumar2311**
Fixes #7630

## Context
Previous PR #9353 attempted these checks but went stale due to metadata 
formatting issues. This PR implements the same checks with correct metadata.

## Changes
Added 3 new Kubernetes workload security checks:

1. `core_memory_limits_set` — Ensure memory limits are set on all containers
2. `core_memory_requests_set` — Ensure memory requests are set on all containers  
3. `core_image_tag_fixed` — Ensure image tag is not set to Latest or Blank

## Metadata fixes from PR #9353 review
- ✅ Description, Risk, Recommendation use Markdown formatting
- ✅ RelatedUrl is empty for Kubernetes checks
- ✅ CheckType is empty for Kubernetes checks
- ✅ No hub links in Recommendation URL

## Testing
Unit tests included for all 3 checks covering PASS, FAIL, and no-pods scenarios.

## Summary by CodeRabbit

* **New Features**
  * Added Kubernetes core checks for container security/resilience: fixed image tags, CPU/memory requests, CPU/memory limits, and configured liveness/readiness probes.
  * Updated the unreleased changelog to include the new “Kubernetes core checks” coverage.
* **Documentation**
  * Added full check documentation and remediation guidance for the new Kubernetes core checks.
* **Tests**
  * Added unit tests for pass/fail scenarios and edge cases (including `latest`/missing tags, digest images, missing/empty resource settings, and ignoring init/ephemeral containers).
* **Refactor**
  * Improved internal Kubernetes pod container handling for consistent evaluation across container types.

### CODE REVIEW

**Comment by HugoPBrito** [comment]
Hi @Nikhilkumar2311!

Thank you a lot for this work.

I'll continue bringing changes from the recently closed #9353 in order to include all check code and have this merged as soon as possible.

File: prowler/providers/kubernetes/services/core/core_image_tag_fixed/core_image_tag_fixed.metadata.json [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+      "Terraform": "```hcl\nresource \"kubernetes_pod\" \"<example_resource_name>\" {\n  metadata { name = \"<example_resource_name>\" }\n  spec {\n    container {\n      name  = \"<example_resource_name>\"\n      image = \"nginx:1.25.3\" # Critical: use a fixed version tag, never 'latest' or untagged\n    }\n  }\n}\n```"
+    },
+    "Recommendation": {
+      "Text": "Pin all container images to a **specific version tag** or **digest** (`@sha256:...`) to ensure reproducible, auditable deployments.\n\n- Avoid `latest` and untagged images in production\n- Use **image digests** for maximum immutability\n- Enforce tag policies with **admission controllers** (e.g., OPA Gatekeeper, Kyverno)\n- Integrate image scanning into CI/CD pipelines for **defense in depth**",
+      "Url": "https://hub.prowler.com/check/core_image_tag_fixed"
+    }
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_image_tag_fixed/core_image_tag_fixed.metadata.json:29]: Fixed the actionable metadata issue in `a7cbcbb`: `Remediation.Code.CLI` is now non-empty.

I am intentionally not adding a `Compliance` block or replacing `Remediation.Recommendation.Url`. Nearby Kubernetes metadata does not use per-check `Compliance` blocks, and `CheckMetadata` currently validates that `Recommendation.Url` must point to `https://hub.prowler.com/...`. Direct Kubernetes documentation remains in `AdditionalURLs`.

File: prowler/providers/kubernetes/services/core/core_memory_limits_set/core_memory_limits_set.metadata.json [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    "https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/"
+  ],
+  "Remediation": {
+    "Code": {
+      "CLI": "",
+      "NativeIaC": "",
+      "Other": "1. Open your Kubernetes UI (e.g., cloud provider console or Dashboard)\n2. Navigate to the failing workload (Deployment/StatefulSet/DaemonSet) or Pod and choose Edit YAML/Manifest\n3. Set resources.limits.memory for each affected container:\n   - For controllers: spec.template.spec.containers[].resources.limits.memory\n   - For standalone Pods: spec.containers[].resources.limits.memory\n4. Save changes (controllers will roll out new Pods). If it is a standalone Pod, delete and recreate it to apply the change",
+      "Terraform": "```hcl\nresource \"kubernetes_pod\" \"<example_resource_name>\" {\n  metadata { name = \"<example_resource_name>\" }\n  spec {\n    container {\n      name  = \"<example_resource_name>\"\n      image = \"<IMAGE>\"\n      resources {\n        limits = {\n          memory = \"128Mi\" # Critical: sets a memory ceiling to prevent unbounded consumption\n        }\n      }\n    }\n  }\n}\n```"
+    },
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_memory_limits_set/core_memory_limits_set.metadata.json:25]: Fixed the actionable part in `a7cbcbb`: `Remediation.Code.CLI` is now populated.

I am not adding a `Compliance` block because nearby Kubernetes check metadata does not use that structure and the current metadata model does not require it for these checks.

File: prowler/providers/kubernetes/services/core/core_memory_limits_set/core_memory_limits_set.metadata.json [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+    "Recommendation": {
+      "Text": "Set **memory limits** on every container to prevent unbounded consumption and protect node stability.\n\n- Start with observed usage plus headroom; tune with **VPA** recommendations\n- Combine with **LimitRanges** and **ResourceQuotas** at the namespace level for **defense in depth**\n- Monitor memory usage and OOMKill events to right-size limits over time",
+      "Url": "https://hub.prowler.com/check/core_memory_limits_set"
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_memory_limits_set/core_memory_limits_set.metadata.json:28]: Won’t change this one. `CheckMetadata` validates that `Remediation.Recommendation.Url` must point to `https://hub.prowler.com/...`, so replacing it with a direct Kubernetes documentation URL would break the project metadata contract.

The direct Kubernetes documentation link stays in `AdditionalURLs`, which is the correct place for it.

File: prowler/providers/kubernetes/services/core/core_memory_requests_set/core_memory_requests_set.metadata.json [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    "https://kubernetes.io/docs/tasks/administer-cluster/manage-resources/memory-default-namespace/"
+  ],
+  "Remediation": {
+    "Code": {
+      "CLI": "",
+      "NativeIaC": "",
+      "Other": "1. Open your Kubernetes UI (e.g., cloud provider console or Dashboard)\n2. Navigate to the failing workload (Deployment/StatefulSet/DaemonSet) or Pod and choose Edit YAML/Manifest\n3. Set resources.requests.memory for each affected container:\n   - For controllers: spec.template.spec.containers[].resources.requests.memory\n   - For standalone Pods: spec.containers[].resources.requests.memory\n4. Save changes (controllers will roll out new Pods). If it is a standalone Pod, delete and recreate it to apply the change",
+      "Terraform": "```hcl\nresource \"kubernetes_pod\" \"<example_resource_name>\" {\n  metadata { name = \"<example_resource_name>\" }\n  spec {\n    container {\n      name  = \"<example_resource_name>\"\n      image = \"<IMAGE>\"\n      resources {\n        requests = {\n          memory = \"64Mi\" # Critical: guarantees memory for the scheduler to make informed placement\n        }\n      }\n    }\n  }\n}\n```"
+    },
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_memory_requests_set/core_memory_requests_set.metadata.json:25]: Fixed the actionable part in `a7cbcbb`: `Remediation.Code.CLI` is now populated.

I am not adding a `Compliance` block because nearby Kubernetes check metadata does not use that structure and the current metadata model does not require it for these checks.

File: prowler/providers/kubernetes/services/core/core_memory_requests_set/core_memory_requests_set.metadata.json [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    "Recommendation": {
+      "Text": "Set **memory requests** on every container so the scheduler can guarantee memory allocation and place pods effectively.\n\n- Base requests on observed steady-state usage; use **VPA** for data-driven sizing\n- Enforce minimum requests via **LimitRanges** at the namespace level\n- Combine with **ResourceQuotas** for **defense in depth** against overcommitment",
+      "Url": "https://hub.prowler.com/check/core_memory_requests_set"
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_memory_requests_set/core_memory_requests_set.metadata.json:28]: Won’t change this one. `CheckMetadata` validates that `Remediation.Recommendation.Url` must point to `https://hub.prowler.com/...`, so replacing it with a direct Kubernetes documentation URL would break the project metadata contract.

The direct Kubernetes documentation link stays in `AdditionalURLs`, which is the correct place for it.

File: prowler/providers/kubernetes/services/core/core_cpu_limits_set/core_cpu_limits_set.py [hunk 6]
```diff
@@ -0,0 +1,26 @@
+from prowler.lib.check.models import Check, Check_Report_Kubernetes
+from prowler.providers.kubernetes.services.core.core_client import core_client
+
+
+class core_cpu_limits_set(Check):
+    def execute(self) -> list[Check_Report_Kubernetes]:
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_cpu_limits_set/core_cpu_limits_set.py:8]: Fixed in `d92753d`: added concise class and `execute()` docstrings.

File: prowler/providers/kubernetes/services/core/core_cpu_requests_set/core_cpu_requests_set.py [hunk 7]
```diff
@@ -0,0 +1,28 @@
+from prowler.lib.check.models import Check, Check_Report_Kubernetes
+from prowler.providers.kubernetes.services.core.core_client import core_client
+
+
+class core_cpu_requests_set(Check):
+    def execute(self) -> list[Check_Report_Kubernetes]:
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_cpu_requests_set/core_cpu_requests_set.py:8]: Fixed in `d92753d`: added concise class and `execute()` docstrings.

File: prowler/providers/kubernetes/services/core/core_liveness_probe_configured/core_liveness_probe_configured.metadata.json [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+  "CheckID": "core_liveness_probe_configured",
+  "CheckTitle": "Pod containers have liveness probes configured",
+  "CheckType": [],
+  "ServiceName": "core",
+  "SubServiceName": "",
+  "ResourceIdTemplate": "",
+  "Severity": "medium",
+  "ResourceType": "KubernetesPod",
+  "Description": "Ensure each container has a liveness probe configured to detect and restart unhealthy containers.",
+  "Risk": "Without liveness probes, failed containers may remain running causing degraded service or resource waste.",
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_liveness_probe_configured/core_liveness_probe_configured.metadata.json:12]: Fixed in `d92753d`: clarified the metadata description to say regular pod containers.

File: prowler/providers/kubernetes/services/core/core_liveness_probe_configured/core_liveness_probe_configured.py [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+
+            for container in (pod.containers or {}).values():
+                if not container.liveness_probe:
+                    report.status = "FAIL"
+                    report.status_extended = f"Pod {pod.name} container {container.name} does not have a liveness probe configured."
+                    break
+
+            findings.append(report)
+
+        return findings
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_liveness_probe_configured/core_liveness_probe_configured.py:28]: Fixed in `d92753d`: added concise class and `execute()` docstrings.

File: prowler/providers/kubernetes/services/core/core_readiness_probe_configured/core_readiness_probe_configured.metadata.json [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  "CheckID": "core_readiness_probe_configured",
+  "CheckTitle": "Pod containers have readiness probes configured",
+  "CheckType": [],
+  "ServiceName": "core",
+  "SubServiceName": "",
+  "ResourceIdTemplate": "",
+  "Severity": "medium",
+  "ResourceType": "KubernetesPod",
+  "Description": "Ensure each container has a readiness probe configured to signal when it is ready to serve traffic.",
+  "Risk": "Without readiness probes, services may receive traffic before containers are ready, causing errors.",
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_readiness_probe_configured/core_readiness_probe_configured.metadata.json]: Fixed in `d92753d`: clarified the title and description to say regular pod containers.

File: prowler/providers/kubernetes/services/core/core_readiness_probe_configured/core_readiness_probe_configured.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+
+            for container in (pod.containers or {}).values():
+                if not container.readiness_probe:
+                    report.status = "FAIL"
+                    report.status_extended = f"Pod {pod.name} container {container.name} does not have a readiness probe configured."
+                    break
+
+            findings.append(report)
+
+        return findings
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_readiness_probe_configured/core_readiness_probe_configured.py:28]: Fixed in `d92753d`: added concise class and `execute()` docstrings.

File: prowler/providers/kubernetes/services/core/core_image_tag_fixed/core_image_tag_fixed.py [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+            report.status_extended = (
+                f"Pod {pod.name} has fixed image tags on all containers."
+            )
+
+            for container in (pod.containers or {}).values():
+                image = container.image
+                if not _has_fixed_image_tag(image):
+                    report.status = "FAIL"
+                    report.status_extended = f"Pod {pod.name} has container {container.name} with image '{image}' that does not use a fixed tag."
+                    break
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_image_tag_fixed/core_image_tag_fixed.py]: Fixed in `edd35bf`: `core_image_tag_fixed` now validates regular, init, and ephemeral containers, with tests for init `latest`, ephemeral missing tag, and fixed init/ephemeral containers without regular containers.

File: prowler/providers/kubernetes/services/core/core_memory_limits_set/core_memory_limits_set.py [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+
+            for container in (pod.containers or {}).values():
+                if (
+                    not container.resources
+                    or not container.resources.get("limits")
+                    or not container.resources["limits"].get("memory")
+                ):
+                    report.status = "FAIL"
+                    report.status_extended = f"Pod {pod.name} does not have memory limits set on container {container.name}."
+                    break
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_memory_limits_set/core_memory_limits_set.py:35]: Fixed in `edd35bf`: memory limits now use the same defensive `isinstance(..., dict)` guard pattern as the CPU checks.

File: prowler/providers/kubernetes/services/core/core_memory_requests_set/core_memory_requests_set.py [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+
+            for container in (pod.containers or {}).values():
+                if (
+                    not container.resources
+                    or not container.resources.get("requests")
+                    or not container.resources["requests"].get("memory")
+                ):
+                    report.status = "FAIL"
+                    report.status_extended = f"Pod {pod.name} does not have memory requests set on container {container.name}."
+                    break
```
[the remark this answers is not in the document]
Comment (HugoPBrito) [reply prowler/providers/kubernetes/services/core/core_memory_requests_set/core_memory_requests_set.py:35]: Fixed in `edd35bf`: memory requests now use the same defensive `isinstance(..., dict)` guard pattern as the CPU checks.

**Review by HugoPBrito — APPROVED** [review]

**Comment by HugoPBrito** [comment]
Thanks a lot for your contribution @Nikhilkumar2311! 🚀
